### PR TITLE
Support fuzzy patch

### LINKED ISSUE #5: [bug] conan.patch incorrectly applied
initial issue: https://github.com/conan-io/conan-center-index/issues/214, https://github.com/conan-io/conan/issues/5963
it turns out conan's `tools.patch` method applies patches incorrectly after the switch to `patch_ng` package. e.g. its behavior significantly differs from the `patch` command, or `git apply` / `git am`

test case to reproduce:
https://github.com/conan-io/conan/compare/develop...SSE4:patch_issue?expand=1#diff-11632e511bcb65a506566fb2157b051eR27
reduced (short) test case:
https://github.com/conan-io/conan/compare/develop...SSE4:patch_issue?expand=1#diff-11632e511bcb65a506566fb2157b051eR409

in terminal:
```
git clone git@github.com:boostorg/boost.git
cd boost
git checkout boost-1.70.0
curl https://raw.githubusercontent.com/conan-io/conan-center-index/5d3a76231eeab86d24dda08cb0ee92fa307ef50f/recipes/boost/all/patches/bcp_namespace_issues.patch > 1.patch
patch -p1 < 1.patch
cat Jamroot | grep "1\.70\.0"
cat Jamroot | grep "1\.71\.0"
```
there is no `1.71.0` in the output (but in case of conan, it is)

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10 Enterprise
  * Compiler+version: N/A
  * Conan version: 1.19.2
  * Python version: 3.7.3

### PR DESCRIPTION
**Pull request by uilianries**
fixes #5 

Should close https://github.com/conan-io/conan/issues/5963 too

### CODE REVIEW

**Comment by uilianries** [comment]
I see. I think it's because the hunk `@@ -1,3 +1,4 @@`

The original file starts at line 1 and is 3 lines size. The initial validation didn't detect the my patch because it only checked the first 3 lines, ignoring `Z`.
In your case, `YYYY` comes into the original hunk.

We need to change the patch applying, but also the initial validation. The question is, should we add an option to allow fuzzy patch? We are breakingthe original behavior with this patch.

**Comment by jgsogo** [comment]
From my point of view:
 1. **FIX:** a fuzzy-patch after the actual changes in the patch-file should fail (as it is failing if the fuzzy is before the actual changes). If we had had this behavior before, the issue with boost wouldn't have happened.
 2. FEATURE: add an argument/option to allow fuzzy patches. But this feature could wait.

**Comment by uilianries** [comment]
Fair enough! I'll check the bug again. Thanks for reporting this new case.

File: patch_ng.py [hunk 1]
```diff
@@ -1050,7 +1044,11 @@ def apply(self, strip=0, root=None):
 
         # check hunks in source file
         if lineno+1 < hunk.startsrc+len(hunkfind)-1:
-          if line.rstrip(b"\r\n") == hunkfind[hunklineno]:
+          raw_line = line.rstrip(b"\r\n")
+          patch_line = hunkfind[hunklineno]
+          if (raw_line == patch_line) or \
+             (fuzz and (not raw_line.startswith(b"+") and not patch_line.startswith(b"+")) and
+             (not raw_line.startswith(b"-") and not patch_line.startswith(b"-"))):
```
Comment (jgsogo) [inline patch_ng.py]: What's the motivation for `raw_line.startswith(b"+")` or `raw_line.startswith(b"-")`? If `line` is the line from the original file, then I don't see any reason for those checks.

Comment (uilianries) [reply patch_ng.py]: nope, I've removed on latest patch. good eye! ;)

File: tests/run_tests.py [hunk 2]
```diff
@@ -424,6 +425,13 @@ def test_delete_file(self):
         pto.apply(strip=0, root=treeroot)
         self.assertFalse(os.path.exists(os.path.join(treeroot, 'deleted')))
 
+    def test_fuzzy_patch(self):
+        treeroot = join(self.tmpdir, 'rootparent')
+        shutil.copytree(join(TESTS, '10fuzzy'), treeroot)
+        pto = patch_ng.fromfile(join(TESTS, '10fuzzy/10fuzzy.patch'))
+        self.assertTrue(pto.apply(root=treeroot, fuzz=True))
+        self.assertFalse(pto.apply(root=treeroot, fuzz=False))
```
[the remark this answers is not in the document]
Comment (uilianries) [reply tests/run_tests.py]: Yeah, it sounds much neater! I gonna improve it, thanks

[the remark this answers is not in the document]
Comment (uilianries) [reply tests/run_tests.py]: @jgsogo Done, please take a look

**Comment by uilianries** [comment]
@jgsogo Thanks for reviewing!! :muscle:

**Comment by lasote** [comment]
Released `patch-ng==1.17.2`

**Comment by uilianries** [comment]
thanks! I gonna restart that PR in Conan

**Pull request merged**